### ISSUE TITLE
org.primefaces.model.UploadedFileWrapper throws NotSerializableException

### ISSUE THREAD

**Issue by suewonjp** [primefaces/primefaces#1805]
## 1) Environment
- PrimeFaces version: 
  - from 5.3.RC1 and all later versions
- Does it work on the newest PrimeFaces version?
  - NO
- Application server + version:
  - Tomcat 8.0
  - myfaces 2.2.8
  - Springframework 3.2.8.RELEASE
  - Spring web flow 2.4.2.RELEASE
- Affected browsers:
  - All
## 2) Expected behavior

File upload should work as expected
## 3) Actual behavior

NotSerializableException occurs
## 4) Steps to reproduce

Simple file upload.
## 5) Sample XHTML

Sorry, preparing the sample was kind of difficult, but I've found the problematic spot and have suggested a solution. Please read the following discussion.
## 6) Sample bean

N/A

---

When I tried to upgrade Primefaces 5.2 -> 5.3 (No other code touched, only Primefaces upgrade) for my project, I've met this issue.

This problem might occur on code with Spring Web Flow, but that is just a guess...

After some track-down, I've found the specific commit _**972eac8**_  causes the problem in my project.

By making UploadedFileWrapper class implement Serializable interface, the exception was gone.

```
-public class UploadedFileWrapper implements UploadedFile, FacesWrapper<UploadedFile>, StateHolder {
+public class UploadedFileWrapper implements UploadedFile, FacesWrapper<UploadedFile>, StateHolder, Serializable {
```

I'm not sure it's a correct solution, but shouldn't UploadedFileWrapper implement Serializable  interface?

**Comment by tandraschko** [comment]
Reverted it as it won't work. Neither Part or FileItem (childs of Native/DefaultUploadedFile)

**Comment by bsanders1979** [comment]
I realize this is an old ticket, but I just upgraded from 5.1 to 11 (huge jump, I know) and I'm getting this error when trying to upload a file. We tried it a few different ways and can't get it to work. Tried using both native and commons upload. When we go native, org.primefaces.component.fileupload.CommonsFileUploadDecoder.getRequest(FacesContext) returns null because "request instanceof MultipartRequest" condition is never met. Do you have any suggestions? Thanks.
* Tomcat 8.5
* Java 8
* JSF 2.2

**Comment by melloware** [comment]
Can you post your whole stack trace?  I don't think you can use native because you are still on JSF 2.2 so are you using Apache Commons FileUpload 1.4 in your war?

Have your filter and web.xml values set for Commons FileUpload?

**Comment by bsanders1979** [comment]
According to the documentation, JSF 2.2 is the bare minimum.
> native: Native mode uses servlet 3.x Part API to upload the files, and if JSF runtime is less than 2.2, an exception is being thrown.

This is with
* primefaces.UPLOADER=commons
* filter/mapping for PrimeFaces FileUpload Filter
* Tomcat 8.5.65
* JSF 2.2.20
* commons-fileupload 1.4
* commons-io 2.2
* PF 11
```
javax.faces.FacesException: java.io.NotSerializableException: org.primefaces.model.file.UploadedFileWrapper
	at com.sun.faces.renderkit.ResponseStateManagerImpl.getViewState(ResponseStateManagerImpl.java:141)
	at javax.faces.application.StateManager.getViewState(StateManager.java:594)
	at com.sun.faces.context.PartialViewContextImpl.renderState(PartialViewContextImpl.java:483)
	at com.sun.faces.context.PartialViewContextImpl.processPartial(PartialViewContextImpl.java:325)
	at org.primefaces.context.PrimePartialViewContext.processPartial(PrimePartialViewContext.java:65)
	at javax.faces.context.PartialViewContextWrapper.processPartial(PartialViewContextWrapper.java:219)
	at javax.faces.component.UIViewRoot.encodeChildren(UIViewRoot.java:1004)
	at javax.faces.component.UIComponent.encodeAll(UIComponent.java:1856)
	at com.sun.faces.application.view.FaceletViewHandlingStrategy.renderView(FaceletViewHandlingStrategy.java:430)
	at com.sun.faces.application.view.MultiViewHandler.renderView(MultiViewHandler.java:134)
	at javax.faces.application.ViewHandlerWrapper.renderView(ViewHandlerWrapper.java:337)
	at com.sun.faces.lifecycle.RenderResponsePhase.execute(RenderResponsePhase.java:120)
	at com.sun.faces.lifecycle.Phase.doPhase(Phase.java:101)
	at com.sun.faces.lifecycle.LifecycleImpl.render(LifecycleImpl.java:219)
	at javax.faces.webapp.FacesServlet.service(FacesServlet.java:659)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at com.[redacted].pm.filter.DeviceIdCookieFilter.doFilter(DeviceIdCookieFilter.java:38)
	at org.omnifaces.filter.HttpFilter.doFilter(HttpFilter.java:77)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.omnifaces.filter.CacheControlFilter.doFilter(CacheControlFilter.java:226)
	at org.omnifaces.filter.HttpFilter.doFilter(HttpFilter.java:77)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.primefaces.webapp.filter.FileUploadFilter.doFilter(FileUploadFilter.java:108)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:317)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:127)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:91)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:114)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.session.SessionManagementFilter.doFilter(SessionManagementFilter.java:137)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.authentication.AnonymousAuthenticationFilter.doFilter(AnonymousAuthenticationFilter.java:111)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:170)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.savedrequest.RequestCacheAwareFilter.doFilter(RequestCacheAwareFilter.java:63)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:200)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:116)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.header.HeaderWriterFilter.doFilterInternal(HeaderWriterFilter.java:64)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter.doFilterInternal(WebAsyncManagerIntegrationFilter.java:56)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:105)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:214)
	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:177)
	at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:346)
	at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:262)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.orm.hibernate4.support.OpenSessionInViewFilter.doFilterInternal(OpenSessionInViewFilter.java:151)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:201)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:544)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:143)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:81)
	at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:698)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:364)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:616)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:831)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1629)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.NotSerializableException: org.primefaces.model.file.UploadedFileWrapper
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1184)
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:348)
	at java.util.ArrayList.writeObject(ArrayList.java:766)
	at sun.reflect.GeneratedMethodAccessor919.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:1140)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1496)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
	at java.io.ObjectOutputStream.writeArray(ObjectOutputStream.java:1378)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1174)
	at java.io.ObjectOutputStream.writeArray(ObjectOutputStream.java:1378)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1174)
	at java.io.ObjectOutputStream.writeArray(ObjectOutputStream.java:1378)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1174)
	at java.io.ObjectOutputStream.writeArray(ObjectOutputStream.java:1378)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1174)
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:348)
	at java.util.HashMap.internalWriteEntries(HashMap.java:1790)
	at java.util.HashMap.writeObject(HashMap.java:1363)
	at sun.reflect.GeneratedMethodAccessor920.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:1140)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1496)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:348)
	at com.sun.faces.renderkit.ClientSideStateHelper.doWriteState(ClientSideStateHelper.java:425)
	at com.sun.faces.renderkit.ClientSideStateHelper.writeState(ClientSideStateHelper.java:187)
	at com.sun.faces.renderkit.ResponseStateManagerImpl.getViewState(ResponseStateManagerImpl.java:139)
	... 83 more
```

**Comment by melloware** [comment]
Please provide an executable example using the [PrimeFaces Test](https://github.com/primefaces/primefaces-test) project.  It is the only way developers can debug your problem to help.

**Comment by bsanders1979** [comment]
I will work on putting something together. What’s interesting is that not all instances are affected. But, this particular instance has a hierarchy along the lines of TabView -> Wizard -> DataTable -> RowExpander. Saw this in the response payload. Not sure if it offers any valuable clues or not.

![image](https://user-images.githubusercontent.com/1124276/170115598-bb538e87-7002-4a2e-b74d-ea59f3be805f.png)

**Comment by melloware** [comment]
Yep sample will help because I can't reproduce in my environment.

**Comment by bsanders1979** [comment]
Ok, I figured it out. State Saving is set to client and it happens when you upload a file and the component is inside of a table row. I suppose I could set the state saving to server, but not sure of the impact of doing so.

mvn clean jetty:run -Pmojarra22 -Djetty.port=5000

[FileUpload-NotSerializable.zip](https://github.com/primefaces/primefaces/files/8886117/FileUpload-NotSerializable.zip)

**Comment by melloware** [comment]
Thanks for the reproducer!  I have reopened the Issue to investigate.

**Comment by melloware** [comment]
@bsanders1979 when I run your reproducer if fails with Mojarra but works with MyFaces.

`mvn clean jetty:run -Pmyfaces23` you will see it works and throws no exception.  So this looks like a Mojarra bug.

Reported here: https://github.com/eclipse-ee4j/mojarra/issues/5117

**Comment by bsanders1979** [comment]
Thanks for the update. I'm not opposed to switching implementations or maybe even switching from client to server. Just not sure if there are any gotchas I need to be aware of.

**Comment by melloware** [comment]
I would say go to MyFaces if you have that option. It's rock solid, faster, and much less gotcha bugs like this one.

**Comment by dmkaw** [comment]
@melloware can you please check once again if its really working on MyFaces?
I have run test suite provided by @bsanders1979 with myfaces23 profile and it is indeed not showing error in console, but error is still being thrown, you can see it in browser log, or by catching exception in debugger.

**Comment by melloware** [comment]
I have reopened the issue.  You are right MyFaces is not throwing the exception but the error is in the browser response.

**Comment by melloware** [comment]
ahhh @tandraschko already debugged this issue.

The problem is with Native JSF uploading its using Servlet 3.0 to do the upload instead of Apache Commons Upload.

See: https://github.com/primefaces/primefaces/blob/dc062460b17ac3564118b78e1e44eae2467a05c5/primefaces/src/main/java/org/primefaces/model/file/NativeUploadedFile.java#L42

The `NativeUploadedFile.java` stores  a `javax.servlet.http.Part` which is NOT Serializable and thus the error.  I could make that `Part` transient but then it really hasn't serialized the upload to the client it would only make the error go away.

I think that is why @tandraschko marked it WontFix.

Apache Commons Upload has the same issue where `private FileItem fileItem;` is not Serializable.

https://github.com/primefaces/primefaces/blob/dc062460b17ac3564118b78e1e44eae2467a05c5/primefaces/src/main/java/org/primefaces/model/file/CommonsUploadedFile.java#L43

**Comment by tandraschko** [comment]
in theory should JSF impl call methods from UploadedFileWrapper#saveState and not serialize the object
the value/uploaded file is then of course only available in the current reqeust and should not be saved

maybe its not used because object nesting or something? i think someone need to debug

**Comment by melloware** [comment]
Let me see if I can debug.

**Comment by melloware** [comment]
OK I am not sure whats going on here but I don't see it calling "SaveState" or "isTransient" of the UploadFilesWrapper its just trying to serialize it in `ClientSideStateHelper.java`  I would think this object would be ignored because its returning `transient` true and eveyrthing works in mode `server` just not in mode `client` on either MyFaces or Mojarra.

**Comment by tandraschko** [comment]
maybe we should just make it transient if both are by spec not serializable

**Comment by melloware** [comment]
When I enable `com.sun.faces.enableClientStateDebugging` I get this output.

```
javax.faces.FacesException: Serialization error. Path to offending instance: [{j_id1=[Ljava.lang.Object;@98cd835, frmTest:j_idt7=[Ljava.lang.Object;@18379790, 
frmTest:j_idt7:j_idt14=[Ljava.lang.Object;@73500cdf, 
frmTest:j_idt7:j_idt16=[Ljava.lang.Object;@14f6ac0b, j_idt4=[Ljava.lang.Object;@2139a8f5, 
com.sun.faces.DynamicActions=[]}, [Ljava.lang.Object;@18379790, [Ljava.lang.Object;@3af7c624, [Ljava.lang.Object;@8397ad4, [Ljava.lang.Object;@44cb96ec, 
[javax.faces.component.StateHolderSaver@5076a828, javax.faces.component.StateHolderSaver@5bc60457, javax.faces.component.StateHolderSaver@6d3e3979, javax.faces.component.StateHolderSaver@63e3a2e8, javax.faces.component.StateHolderSaver@782fa222, javax.faces.component.StateHolderSaver@69fe06c6, javax.faces.component.StateHolderSaver@3d83adf6, javax.faces.component.StateHolderSaver@130022dc, javax.faces.component.StateHolderSaver@4abee9c6, javax.faces.component.StateHolderSaver@5c047a54, javax.faces.component.StateHolderSaver@1bea1871, javax.faces.component.StateHolderSaver@2f34d6ec, javax.faces.component.StateHolderSaver@7bc38330, javax.faces.component.StateHolderSaver@4c69e797, javax.faces.component.StateHolderSaver@12aa6a5a, javax.faces.component.StateHolderSaver@4a17894c, javax.faces.component.StateHolderSaver@439fce84, javax.faces.component.StateHolderSaver@70e50cdd, javax.faces.component.StateHolderSaver@7bc2c4d5, javax.faces.component.StateHolderSaver@50e19682, javax.faces.component.StateHolderSaver@5d57fce0, javax.faces.component.StateHolderSaver@5ebe1b34, javax.faces.component.StateHolderSaver@7f636179, javax.faces.component.StateHolderSaver@6de2a6e1, javax.faces.component.StateHolderSaver@1d3d844e, javax.faces.component.StateHolderSaver@7d33ed24, javax.faces.component.StateHolderSaver@7f4af6de, javax.faces.component.StateHolderSaver@57da4893, javax.faces.component.StateHolderSaver@290acf1, javax.faces.component.StateHolderSaver@44aeeb2f, javax.faces.component.StateHolderSaver@48be33fb, javax.faces.component.StateHolderSaver@174968f6, javax.faces.component.StateHolderSaver@1669f467, javax.faces.component.StateHolderSaver@6bc30d2, javax.faces.component.StateHolderSaver@79a4fc42, javax.faces.component.StateHolderSaver@33d705c1, javax.faces.component.StateHolderSaver@7290faed, javax.faces.component.StateHolderSaver@79c1a802, javax.faces.component.StateHolderSaver@623b93a3, javax.faces.component.StateHolderSaver@7e81028a, javax.faces.component.StateHolderSaver@1deca01d, javax.faces.component.StateHolderSaver@40031f10, javax.faces.component.StateHolderSaver@3f2ec4b8, javax.faces.component.StateHolderSaver@572d4849, javax.faces.component.StateHolderSaver@a0117da], javax.faces.component.StateHolderSaver@7f4af6de,
 submittedValue: null value: org.primefaces.model.file.UploadedFileWrapper@6cab8192
 localValueSet: true, org.primefaces.model.file.UploadedFileWrapper@6cab8192]
```

**Comment by bsanders1979** [comment]
FWIW, I decided to set the state saving to server and I think it might actually be an improvement for us. We have some pages with several form tags which increases the payload size/load times with client-side state saving.

**Comment by melloware** [comment]
looks like @tandraschko did fix this issue for 12.0.0 I just tested it and it works with client in 12.0.0-SNAPSHOT.  But yes `server` is most likely the better choice.
